packages/files/utils.js

```diff
@@ -0,0 +1,388 @@
+            let fileError
+            const capacitor = new WriteStream()
+
+            capacitor.on('error', () => {
+                stream.unpipe()
+                stream.resume()
+            })
+
+            stream.on('limit', () => {
+                stream.unpipe()
+                capacitor.destroy(fileError)
+                sendError(res, 413, `File truncated as it exceeds the ${processRequestOptions.maxFileSize} byte size limit.`)
+            })
+
+            stream.on('error', (error) => {
+                fileError = error
+                stream.unpipe()
+                capacitor.destroy(fileError)
+            })
+
+            const file = {
+                filename,
+                mimetype,
+                encoding,
+                stream,
+                createReadStream (options) {
+                    const error = fileError || (released ? exitError : null)
+                    if (error) throw error
+                    return capacitor.createReadStream(options)
+                },
+                capacitor,
+            }
+
+            Object.defineProperty(file, 'capacitor', {
+                enumerable: false,
+                configurable: false,
+                writable: false,
+            })
+
+            stream.pipe(capacitor)
+            if (!Array.isArray(req.files)) req.files = [file]
+            else req.files.push(file)
+        })
+
+        parser.once('filesLimit', () =>
+            exit(() => sendError(res, 413, `${processRequestOptions.maxFiles} max file uploads exceeded.`))
+        )
+
+        parser.on('field', (fieldName, value, { valueTruncated }) => {
+            if (valueTruncated) {
+                return exit(() =>
+                    sendError(res, 413, `The ‘${fieldName}’ multipart field value exceeds the ${processRequestOptions.maxFieldSize} byte size limit.`)
+                )
+            }
+
+            if (fieldName === 'meta') {
+                meta = parseAndValidateMeta(value, req, res, exit)
+            }
+
+            req.meta = meta
+        })
+
+        parser.once('finish', () => {
+            req.unpipe(parser)
+            req.resume()
+
+            if (!Array.isArray(req.files)) {
+                return exit(() => sendError(res, 400, 'Missing attached files'))
+            }
+            if (!meta) {
+                return exit(() => sendError(res, 400, 'Missing multipart field "meta"'))
+            }
+            next()
+        })
+
+        req.once('close', () => {
+            if (!req.readableEnded) {
+                exit(() => sendError(res, 499, 'Request disconnected during file upload stream parsing.'))
+            }
+        })
+
+        res.once('close', () => { released = true })
+        req.pipe(parser)
+    }
+}
+
+const fileStorageHandler = ({ keystone, appClients }) => {
+    return async function (req, res) {
+        const { meta, files } = req
```

Comment:
Maybe more scecific field on req instead of meta

Reply:
const key = Symbol('fileMeta')
req[key]

[[['fileMeta']]]

Reply:
This is done